### ISSUE TITLE
Using GStreamer on a Raspberry Pi to stream SRT: Failed to set SRTO_LINGER

### ISSUE THREAD

**Issue by HR-KSSK** [Haivision/srt#1374]
The following question is about GStreamer and SRT - but I think it is more SRT-related so I am posting it over here:

I am trying to run a GStreamaer Pipeline to stream a V4L2 Device with a Raspberry Pi 4 B (4GB). For this I am using the following Pipeline from the SRT-Docs:
```
GST_DEBUG=2 gst-launch-1.0 v4l2src ! videoconvert ! x264enc bitrate=8000 tune=zerolatency speed-prese
t=superfast byte-stream=true threads=1 key-int-max=15 intra-refresh=true ! video/x-h264, profile=baseline ! mpegtsmux !
srtserversink uri=srt://:8888/ latency=100
```

The problem is, that the streaming does not start and throws these error logs:
```
0:00:00.003756396  5333   0x74e4c0 WARN         GST_PERFORMANCE gstbuffer.c:486:_priv_gst_buffer_initialize: No 64-bit atomic int defined for this platform/toolchain!
Leitung wird auf PAUSIERT gesetzt ...
0:00:00.128904732  5333   0x74e4c0 WARN                 srtsink gstsrtsink.c:160:gst_srt_sink_start:<srtsink0> error: Failed to open SRT: failed to set SRTO_LINGER (reason: Operation not supported: Bad parameters)
0:00:00.129081286  5333   0x74e4c0 WARN                basesink gstbasesink.c:5368:gst_base_sink_change_state:<srtsink0> error: Failed to start
FEHLER: Leitung möchte nicht pausiert werden.
FEHLER: Von Element /GstPipeline:pipeline0/GstSRTSink:srtsink0: Die Ressource konnte nicht zum Schreiben geöffnet werden.
Zusätzliche Fehlerdiagnoseinformation:
gstsrtsink.c(160): gst_srt_sink_start (): /GstPipeline:pipeline0/GstSRTSink:srtsink0:
Failed to open SRT: failed to set SRTO_LINGER (reason: Operation not supported: Bad parameters)
Leitung wird auf NULL gesetzt ...
Leitung wird geleert ...
```
Some information is in German but I can translate it if necessary. I think the problem seems to be that the srtsink can't be opened. Is it possible that this is an issue with the linux firewall? I am not sure where to start solving the problem.. I tried to find some info about the "failed to set SRTO_LINGER" but didn't find any..

I am using GStreamer 1.16.2 which I built with SRT-support according to this repo info:
https://github.com/PietroAvolio/Building-Gstreamer-Raspberry-Pi-With-SRT-Support

Any help appreciated!

**Comment by maxsharabayko** [comment]
Hi @HR-KSSK 
Looks like a small bug in GStreamer plugin.

SRT checks the length of the param provided to `srt_setsockopt(.., len)` when SRTO_LINGER is being set. And this check fails.
```c++
case SRTO_LINGER:
    if (optlen < (int)(sizeof(linger)))
        throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
```

Looking into the code of the [GStreamer plugin](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L137) all socket options are passed as integers. While size of the system structure `linger` must be passed.

As a workaround, you could temporarily delete the [line with linger (gsrtobject.c #L131)](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/ext/srt/gstsrtobject.c#L121) from GStreamer plugin.

TODO:

- [ ] Fix GStreamer to provide linger or don't provide this value (default is 0 for live mode).
- [ ] Fix `SRTO_LINGER` default value in API.md docs.
- [ ] Fix `SRTO_LINGER` setting in SRT (check if 0 is provided as integer) - Does it make sense????
There are several things that could be done here.
A

**Comment by HR-KSSK** [comment]
Hey @maxsharabayko 
thanks for your quick response.

I will try the workaround but as far as I know I have to rebuild GStreamer Bad plugins again by myself, so this will take some time. I will give an update here if this works.

Should I open up an issue at the GStreamer plugin, too?

Greetings!

**Comment by maxsharabayko** [comment]
> Should I open up an issue at the GStreamer plugin, too?

Probably. 
@justinjoy Do you think it makes sense to submit an issue to GStreamer plugin?

**Comment by HR-KSSK** [comment]
Hey again,

I tried your workaround and deleted L121 in gsrtobject.c and rebuilt the bad plugins but the error persists:
```
0:00:00.004380237 32256   0x8a12c0 WARN         GST_PERFORMANCE gstbuffer.c:486:_priv_gst_buffer_initialize: No 64-bit atomic int defined for this platform/toolchain!
Leitung wird auf PAUSIERT gesetzt ...
0:00:00.139982208 32256   0x8a12c0 WARN                 srtsink gstsrtsink.c:160:gst_srt_sink_start:<srtsink0> error: Failed to open SRT: failed to set SRTO_LINGER (reason: Operation not supported: Bad parameters)
0:00:00.140146077 32256   0x8a12c0 WARN                basesink gstbasesink.c:5368:gst_base_sink_change_state:<srtsink0> error: Failed to start
FEHLER: Leitung möchte nicht pausiert werden.
FEHLER: Von Element /GstPipeline:pipeline0/GstSRTSink:srtsink0: Die Ressource konnte nicht zum Schreiben geöffnet werden.
Zusätzliche Fehlerdiagnoseinformation:
gstsrtsink.c(160): gst_srt_sink_start (): /GstPipeline:pipeline0/GstSRTSink:srtsink0:
Failed to open SRT: failed to set SRTO_LINGER (reason: Operation not supported: Bad parameters)
Leitung wird auf NULL gesetzt ...
Leitung wird geleert ...
```

**Comment by justinjoy** [comment]
> > Should I open up an issue at the GStreamer plugin, too?
> 
> Probably.
> @justinjoy Do you think it makes sense to submit an issue to GStreamer plugin?

Yes, it should be. Let's make noise in that community relating to srt. :)
Actually, I am preparing a patch to set or accept all SRT options in GST plugins, rather than creating a wrapper property per option.

**Comment by HR-KSSK** [comment]
Hey @justinjoy  and @maxsharabayko 
I rebuilt the GStreamer Bad Plugin including the simple workaround (delete the line as mentioned above).

Now I can get the pipeline running. The strange thing is that I am not able to view the stream from another client inside the same network via VLC. It just starts buffering but is doing nothing..

I mentioned that the srt uri does not have a host adress in the examples:
`srtserversink uri=srt://:8888`

I also tried:
`srtserversink uri=srt://0.0.0.0:8888`

and:
`srtsink uri=srt://0.0.0.0:8888`

But then I get this error:

```
0:00:00.004242365 18980  0x1d532c0 WARN         GST_PERFORMANCE gstbuffer.c:486:_priv_gst_buffer_initialize: No 64-bit atomic int defined for this platform/toolchain!
Leitung wird auf PAUSIERT gesetzt ...
0:00:00.118435425 18980  0x1d532c0 WARN                    v4l2 gstv4l2object.c:4211:gst_v4l2_object_probe_caps:<v4l2h264enc0:src> Failed to probe pixel aspect ratio with VIDIOC_CROPCAP: Das Argument ist ungültig
Leitung ist aktiv und erfordert keinen VORLAUF …
Leitung wird auf ABSPIELEN gesetzt ...
New clock: GstSystemClock
Verzögerung neu verteilen …
0:00:00.175634010 18980  0x1e4d1b0 WARN          v4l2bufferpool gstv4l2bufferpool.c:810:gst_v4l2_buffer_pool_start:<v4l2h264enc0:pool:src> Uncertain or not enough buffers, enabling copy threshold
0:00:00.189300889 18980 0xb2b029b0 WARN          v4l2bufferpool gstv4l2bufferpool.c:1279:gst_v4l2_buffer_pool_dqbuf:<v4l2h264enc0:pool:src> Driver should never set v4l2_buffer.field to ANY
0:00:00.205944868 18980 0xb2b029b0 WARN            v4l2videoenc gstv4l2videoenc.c:680:gst_v4l2_video_enc_loop:<v4l2h264enc0> Encoder is producing too many buffers
0:00:03.243997290 18980  0x1e41490 WARN               srtobject gstsrtobject.c:1253:gst_srt_object_send_headers:<srtsink0> error: Connection does not exist
FEHLER: Von Element /GstPipeline:pipeline0/GstSRTSink:srtsink0: In die Ressource konnte nicht geschrieben werden.
Zusätzliche Fehlerdiagnoseinformation:
gstsrtobject.c(1253): gst_srt_object_send_headers (): /GstPipeline:pipeline0/GstSRTSink:srtsink0:
Connection does not exist
Execution ended after 0:00:03.123259314
Leitung wird auf PAUSIERT gesetzt ...
Leitung wird auf BEREIT gesetzt ...
16:35:49.952580/gst-launch-1.0*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @126824898
Leitung wird auf NULL gesetzt ...
16:35:49.979232/gst-launch-1.0*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @126824898
16:35:49.979343/gst-launch-1.0*E:SRT.c: IPE: remove_usock: @-1 not found as either socket or group. Removing only from epoll system.
Leitung wird geleert ...
```
The client I use to connect via VLC (3.0.11) is a Windows 10 machine..

Do you have any tips on this? I can post this @GStreamer if this is not the right place for this..

**Comment by justinjoy** [comment]
Actually, `srtserversink` is just an alias of  `srtsink` in GStreamer. 
It looks like your sink failed to connect or accept a connection. How about `srtsink uri=srt://0.0.0.0:8888?mode=listener`?

If not working, that will be good to open an issue to GStreamer gitlab with `GST_DEBUG=srt*:5` log.

**Comment by HR-KSSK** [comment]
Hi again,

that changes the output to:
```
0:00:00.124687729 29070  0x15f1800 DEBUG              srtobject gstsrtobject.c:186:gst_srt_object_new:<GstSRTSink@0x160a2b8> Starting up SRT
0:00:00.125054615 29070  0x15f1800 DEBUG              srtobject gstsrtobject.c:564:gst_srt_object_set_uri:<GstSRTSink@0x160a2b8> set uri to (host: 127.0.0.1, port: 7001) with 0 query strings
0:00:00.125155596 29070  0x15f1800 DEBUG              srtobject gstsrtobject.c:564:gst_srt_object_set_uri:<srtsink0> set uri to (host: 0.0.0.0, port: 8888) with 1 query strings
Leitung wird auf PAUSIERT gesetzt ...
0:00:00.127092211 29070  0x15f1800 DEBUG              srtobject gstsrtobject.c:1001:gst_srt_object_open_full:<srtsink0> Opening SRT socket with parameters: application/x-srt-params, poll-timeout=(int)-1, latency=(int)100, mode=(GstSRTConnectionMode)GST_SRT_CONNECTION_MODE_LISTENER, localaddress=(string)0.0.0.0, localport=(uint)8888;
0:00:00.127299857 29070  0x15f1800 DEBUG              srtobject gstsrtobject.c:744:gst_srt_object_wait_connect:<srtsink0> Binding to 0.0.0.0 (port: 8888)
0:00:00.127647429 29070  0x15f1800 DEBUG              srtobject gstsrtobject.c:759:gst_srt_object_wait_connect:<srtsink0> Starting to listen on bind socket
0:00:00.127826205 29070  0x15d5350 DEBUG              srtobject gstsrtobject.c:624:thread_func:<srtsink0> Waiting a request from caller
Leitung ist aktiv und erfordert keinen VORLAUF …
Leitung wird auf ABSPIELEN gesetzt ...
New clock: GstSystemClock
Verzögerung neu verteilen …
0:00:00.143430826 29070  0x15d5350 DEBUG              srtobject gstsrtobject.c:624:thread_func:<srtsink0> Waiting a request from caller
0:00:00.188487462 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:248:gst_srt_sink_set_caps:<srtsink0> setcaps video/mpegts, systemstream=(boolean)true, packetsize=(int)188
0:00:00.188558591 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:256:gst_srt_sink_set_caps:<srtsink0> 'streamheader' field not present
0:00:00.188584406 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:286:gst_srt_sink_set_caps:<srtsink0> Collected streamheaders: 0 buffers
0:00:00.189029995 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:248:gst_srt_sink_set_caps:<srtsink0> setcaps video/mpegts, systemstream=(boolean)true, packetsize=(int)188, streamheader=(buffer)< 47400030a600ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000b00d0001c100000001e020a2c32941, 474020308b00ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0002b0280001c10000e041f00c050448444d5688040ffffcfc1be041f00a050848444d56ff1b443f5a3175c0 >
0:00:00.189083402 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:264:gst_srt_sink_set_caps:<srtsink0> 'streamheader' field holds array
0:00:00.189116680 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:286:gst_srt_sink_set_caps:<srtsink0> Collected streamheaders: 2 buffers
0:00:00.224243206 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:191:gst_srt_sink_render:<srtsink0> Have streamheaders, ignoring header buffer: 0xb2b05950, pts 0:00:00.031505850, dts 99:99:99.999999999, dur 99:99:99.999999999, size 188, offset none, offset_end none, flags 0x6400
0:00:00.224314521 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:191:gst_srt_sink_render:<srtsink0> Have streamheaders, ignoring header buffer: 0xb2b05b48, pts 0:00:00.031505850, dts 99:99:99.999999999, dur 99:99:99.999999999, size 188, offset none, offset_end none, flags 0x6400
0:00:00.224356409 29070  0x15c9c90 DEBUG                srtsink gstsrtsink.c:191:gst_srt_sink_render:<srtsink0> Have streamheaders, ignoring header buffer: 0xb2b05c98, pts 0:00:00.031505850, dts 99:99:99.999999999, dur 99:99:99.999999999, size 188, offset none, offset_end none, flags 0x2400
0:00:00.234357981 29070  0x15c9c90 DEBUG              srtobject gstsrtobject.c:1081:gst_srt_object_wait_caller:<srtsink0> Waiting connection from caller
0:00:11.357429078 29070  0x15d5350 DEBUG              srtobject gstsrtobject.c:624:thread_func:<srtsink0> Waiting a request from caller
0:00:44.544589599 29070  0x15d5350 DEBUG              srtobject gstsrtobject.c:624:thread_func:<srtsink0> Waiting a request from caller
0:02:12.443820742 29070  0x15d5350 DEBUG              srtobject gstsrtobject.c:624:thread_func:<srtsink0> Waiting a request from caller
0:05:49.870566389 29070  0x15d5350 DEBUG              srtobject gstsrtobject.c:624:thread_func:<srtsink0> Waiting a request from caller
```
which looks promising to me - the problem is that I still cant connect with VLC on another machine on the network...

**Comment by justinjoy** [comment]
If it's not a network issue, it seems to fail to connect or negotiate SRT connection from VLC.
Let me check the reproducible environment before going to GStreamer community.

SRT Listener - RPi
 - GStreamer 1.16.2
 - SRT which version?

SRT Caller
 - VLC 3.0.11 on Windows

Although I haven't noted an exact commit hash, VLC isn't compatible with a specific commit of SRT master.
I am using fdb20ff of SRT to keep compatibility.


Can you confirm which version of SRT are you using?

**Comment by HR-KSSK** [comment]
Hey,
I installed SRT via make and used the latest commit available @GitHub. How can I find out the commit I used?

Inspect shows this but I think that's not what you're asking for:
```
gst-inspect-1.0 srt
Plugin Details:
  Name                     srt
  Description              transfer data via SRT
  Filename                 /usr/local/lib/gstreamer-1.0/libgstsrt.so
  Version                  1.16.2
  License                  LGPL
  Source module            gst-plugins-bad
  Source release date      2019-12-03
  Binary package           GStreamer Bad Plug-ins source release
  Origin URL               Unknown package origin

  srtserversink: SRT sink
  srtclientsink: SRT sink
  srtserversrc: SRT source
  srtclientsrc: SRT source
  srtsink: SRT sink
  srtsrc: SRT source

  6 features:
  +-- 6 elements
```

Greetings

**Comment by justinjoy** [comment]
> I installed SRT via make and used the latest Master available @github. How can I find out the commit I used?

By `git checkout`.

```
$ git clone https://github.com/Haivision/srt.git
$ cd srt
$ git checkout  fdb20ff 
```
Then build and test SRT connection with VLC.

If it works, it will be SRT bug that breaks compatibility in somewhere recent SRT commits.

**Comment by HR-KSSK** [comment]
I can't build SRT with that commit version:
```
Scanning dependencies of target srtsupport_virtual
[  2%] Building CXX object CMakeFiles/srtsupport_virtual.dir/apps/apputil.cpp.o
In file included from /home/pi/srt/new/srt/apps/apputil.cpp:18:
/home/pi/srt/new/srt/apps/apputil.hpp:101:24: error: redefinition of ‘Number StrToNumber(const string&) [with Number = unsigned int; std::__cxx11::string = std::__cxx11::basic_string<char>]’
 template<> inline type StrToNumber(const std::string& s) { return function (s, 0, 0); }
                        ^~~~~~~~~~~
/home/pi/srt/new/srt/apps/apputil.hpp:107:1: note: in expansion of macro ‘STON’
 STON(unsigned int, stoul);
 ^~~~
/home/pi/srt/new/srt/apps/apputil.hpp:101:24: note: ‘Number StrToNumber(const string&) [with Number = unsigned int; std::__cxx11::string = std::__cxx11::basic_string<char>]’ previously declared here
 template<> inline type StrToNumber(const std::string& s) { return function (s, 0, 0); }
                        ^~~~~~~~~~~
/home/pi/srt/new/srt/apps/apputil.hpp:105:1: note: in expansion of macro ‘STON’
 STON(size_t, stoul);
 ^~~~
make[2]: *** [CMakeFiles/srtsupport_virtual.dir/build.make:63: CMakeFiles/srtsupport_virtual.dir/apps/apputil.cpp.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:186: CMakeFiles/srtsupport_virtual.dir/all] Fehler 2
make: *** [Makefile:130: all] Fehler 2
```
I try to read through the code - maybe I can fix it. Do you know any other commit which should work, too? Maybe the build process works with another version.

**Comment by HR-KSSK** [comment]
I tried building with ce07720 and it worked. It is just slightly ahead of the commit u mentioned.

The problem with that is that the output is the same as above. Instead now GStreamer "stops" @
```
Leitung ist vorgelaufen …
Leitung wird auf ABSPIELEN gesetzt ...
0:00:00.313270844 20001  0x20ab350 DEBUG              srtobject gstsrtobject.c:1081:gst_srt_object_wait_caller:<srtsink0> Waiting connection from caller
New clock: GstSystemClock
```
If I call from VLC nothing more happens... I am connected via VPN now - I will give it a try later with direct network access to the Pi.

**Comment by ethouris** [comment]
The `apputil.cpp` file has been newly added at some point. Make sure that you run a clean build. Why don't you simply take the stable release version tagged as `v1.4.1`?

**Comment by justinjoy** [comment]
> I tried building with ce07720 and it worked. It is just slightly ahead of the commit u mentioned.

Then, I suppose that you can build with that commit, but it still doesn't accept the connection. right?
As I remember, somewhere between fdb20ff and ce07720 breaks connection compatibility.

As @ethouris mentioned, how about with `v1.4.1` simply?

**Comment by HR-KSSK** [comment]
Hello again,
after building v1.4.1 I get linking problems with GStreamer now:
```
(gst-launch-1.0:26101): GStreamer-WARNING **: 12:32:05.206: Failed to load plugin '/usr/local/lib/gstreamer-1.0/libgstsrt.so': libsrt.so.1: Can't open Shared-Object: File or Directory not found
WARNING: Corrupted Pipeline: No Element »srtsink«
```
If I inspect GStreamer srt:
```
gst-inspect-1.0 srt
Plugin Details:
  Name                     srt
  Description              transfer data via SRT
  Filename                 /usr/local/lib/gstreamer-1.0/libgstsrt.so
  Version                  1.16.2
  License                  LGPL
  Source module            gst-plugins-bad
  Source release date      2019-12-03
  Binary package           GStreamer Bad Plug-ins source release
  Origin URL               Unknown package origin

  srtsrc: SRT source
  srtsink: SRT sink
  srtclientsrc: SRT source
  srtserversrc: SRT source
  srtclientsink: SRT sink
  srtserversink: SRT sink

  6 features:
  +-- 6 elements
```
So I think there is a problem with a symbolic link which is not linking correctly to the newly created libsrt.so...

Do you know how to fix that? Maybe I just try to reinstall GStreamer Bad-Plugins and try if that makes it work again..

**Comment by justinjoy** [comment]
> after building v1.4.1 I get linking problems with GStreamer now:

Did you build gst-plugins-bad again after back to v1.4.1 of srt?

**Comment by HR-KSSK** [comment]
Hey,

I was not able to get it to work again - even after rebuilding gst-plugins-bad..

Now I managed to get it working again by rebuilding and reinstalling the whole GStreamer 1.16 build including all plugins.

I am able to playback the SRT Stream now! With VMix and VLC player.

But there are some disadvantages:
If I stream the videotestsrt everything works fine and I can connect with multiple clients and watch the stream with several different Apps. But if I stream my v4l2 device with:
`GST_DEBUG=srt*:5 gst-launch-1.0 v4l2src ! "video/x-raw,framerate=50/1,format=UYVY" ! v4l2h264enc extra-controls="controls,video_bitrate=6000000;" ! video/x-h264, profile=high  ! mpegtsmux ! srtsink uri=srt://:9999 latency=100`

I can only connect ONCE with only 1 client and I can't reopen the stream after closing/stopping it in the client app (for example VLC).. But I still get messages in the terminal which show the stream running and an established connection with a client... really strange.

**Comment by justinjoy** [comment]
🤔 `srtsink` is designed to accept multiple callers if it runs as a listener, and I don't have problems making connections repeatedly. 

> Now I managed to get it working again by rebuilding and reinstalling the whole GStreamer 1.16 build including all plugins.

I am curious about this. SRT library should only affect to SRT plugins.

**Comment by HR-KSSK** [comment]
> I am curious about this. SRT library should only affect to SRT plugins.
I think somehow the rebuilding/reinstalling messed up some symlinks on the Pi's system. I don't think that's a problem with the library.

> 🤔 srtsink is designed to accept multiple callers if it runs as a listener, and I don't have problems making connections repeatedly. 
So maybe this is also more Pi-related. Or maybe there is a problem with the way I am getting the stream from my V4L2-Device, bc the testsrc is working. Do you have any experience in streaming v4l2-devices via GStreamer?

**Comment by maxsharabayko** [comment]
Hi @HR-KSSK @justinjoy 
Could you please share a summary of the current status?

> Now I managed to get it working again by rebuilding and reinstalling the whole GStreamer 1.16

Did you use SRT v1.4.1 or the latest master?
 
Does the latest SRT master has connectivity issues to investigate?

**Comment by HR-KSSK** [comment]
Hi @maxsharabayko 

I used a Raspberry Pi 4 B (4GB) as a SRT listener (server).

- using master branch:
No connection possible (tried it with VLC Player and VMix - both on Windows 10 machines)
No difference in using testsrc or "real" v4l2 device as source

- using 1.4.1
Multiple Connections possible when streaming videotestsrc - works as expected
ONE connection possible when streaming v4l2 device as source
-> for any new connection I get correct output in terminal but I can't see the stream on client machines (again VLC and VMix tested)
-> If I cancel the one working connection no new "viewable" connection can be established

If you questions please let me know

**Comment by maxsharabayko** [comment]
Hi @HR-KSSK 
Thanks for the update.
I am mainly concerned about the master branch not able to establish a connection. I don't see any reason for it to anyhow differ from v1.4.1.
I wonder if there is a way to track down this behavior? Might be helpful if you could collect a Wireshark capture of both cases. I assume the caller remains the same in both tests, right?

**Comment by HR-KSSK** [comment]
Hi again,

yes the callers are the same (both windows 10 machines in the same network as the listener).

I will try to collect some info about the network traffic. This will take some time, because upgrading/downgrading srt version seems to be non-trivial...

**Comment by ethouris** [comment]
Closing due to too long inactivity. Please revive if this is still a problem.

**Comment by nh2** [comment]
This issue seems to continue to exist in gstreamer. I get today:

> ```
> Failed to open SRT: failed to set SRTO_LINGER (reason: Operation not supported: Bad parameters)
> ```

The analysis of @maxsharabayko in https://github.com/Haivision/srt/issues/1374#issuecomment-649285791 seems completely on point.

> > Do you think it makes sense to submit an issue to GStreamer plugin?
> 
> Yes, it should be.

It seems that nobody has done this so far.

**Comment by ethouris** [comment]
I understand, but there's really nothing we can do at the SRT side. If the `SRTO_LINGER` option is used wrong way in the GStreamer project, it should be fixed in the GStreamer project.

**Comment by nh2** [comment]
What about

> * [ ] Fix `SRTO_LINGER` default value in API.md docs.

? This suggests that the default value is incorrectly documented. If the default value is to turn linger off, it'd be trivial to PR a deletion of that line into gstreamer.

**Comment by nh2** [comment]
I found that gstreamer already made a fix 5 days ago: https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/1615/diffs

**Comment by maxsharabayko** [comment]
Update. GStreamer fix for `SRTO_LINGER` is [merged](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/1615).

**Comment by lakis1982** [comment]
Hello @maxsharabayko @nh2 @justinjoy @HR-KSSK 

I also have the same problem as @HR-KSSK . In my raspberry pi 3 I installed GStreamer 1.16.3 and SRT following this link https://github.com/PietroAvolio/Building-Gstreamer-Raspberry-Pi-With-SRT-Support (the same link as @HR-KSSK did)

When i give the command to my rapsberry pi 3  gst-launch-1.0 -v rpicamsrc preview=true sensor-mode=5 bitrate=8000000! video/x-h264,width=1640,height=922,framerate=40/1,profile=baseline ! mpegtsmux ! srtsink uri=srt://:8888  I get the following errors:
Setting pipeline to PAUSED ...
ERROR: Pipeline doesn't want to pause.
ERROR: from element /GstPipeline:pipeline0/GstSRTSink:srtsink0: Could not open resource for writing.
Additional debug info:
gstsrtsink.c(160): gst_srt_sink_start (): /GstPipeline:pipeline0/GstSRTSink:srtsink0:
Failed to open SRT: failed to set SRTO_LINGER (reason: Operation not supported: Bad parameters)
Setting pipeline to NULL ...
Freeing pipeline ...

I am not a programmer so I hardly managed to get to this point so far. Could you please let me know how I could fix this problem so that I can send video from my raspberry using the above command ? Should i change something to the above command so that it could work ? 

Any simple help/commands would be really appreciated..

Thank you

Thanks a lot

**Comment by justinjoy** [comment]
At the moment, I should say use `GStreamer >= 1.18` when building from scratch.

**Comment by lakis1982** [comment]
@justinjoy thank you for your quick reply.
As I said I installed the Gstreamer through this guide https://github.com/PietroAvolio/Building-Gstreamer-Raspberry-Pi-With-SRT-Support  by executing the  https://github.com/PietroAvolio/Building-Gstreamer-Raspberry-Pi-With-SRT-Support/blob/master/gstreamer-build.sh  script.
Since my knowledge is at minimum, does this script first uninstalls the existing gstreamer (e.g v 1.16.3) and then installs the new one (e.g. 1.18) as long as I changed the line 4 of this script to BRANCH="1.18" ?? I ask because I see some **sudo make uninstall || true** commands within the script.
